Fixes replace based upgrades

m1kola opened the pull request:
# Description

Fixes replace based upgrades.

Closes #321

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

m1kola commented on internal/resolution/entitysources/catalogdsource.go:
```diff
@@ -110,9 +110,10 @@ func getEntities(ctx context.Context, client client.Client) (input.EntityList, e
-					replacesValue, _ := json.Marshal(replacesProperty{Replaces: b.Replaces})
-					props["olm.replaces"] = string(replacesValue)
+					replacesValue, _ := json.Marshal(replacesProperty{
+						Replaces: fmt.Sprintf("%s-%s%s%s", bundle.Spec.Catalog.Name, b.Replaces, bundle.Spec.Package, ch.Name),
```
So the idea is that: in the entity properties we have formated replaces (vs raw data from catalogd objects) and don't have to format it in `internal/resolution/variablesources/installed_package.go`: we just filter entity source by id of the currently installed entity.

However I'm not entirely sure whether this is what we want to do. This limits replaces to a specific catalog and channel.

awgreene replied:
In OLM V0, which I believe is be the primary usecase for replace/skips/skipRange based upgrades, we supported cross channel updates. I suspect that we'll want to support cross channel upgrades here as well.

m1kola replied:
Yeah, and for [semver upgrade support](https://docs.google.com/document/d/1KIhV8oqr22Bp8lf370Aej2n-3uW_xGcTptONHHPRVYA/edit) we also want:

* The bundle resolved for an operator during an upgrade can come from a different channel or catalog than the currently installed bundle. (i.e. successors can come from any channel and any catalog)
* The bundle resolved for an operator during an upgrade cannot come from a different package than the currently installed bundle.

So I'm looking into other options to solve this.

ankitathomas replied:
I think it may be less restricting to have channel information be a new constraint rather than included into the replaces field keeping cross channel updates in mind, similar to what the upgrade support brief mentions.

perdasilva commented on internal/resolution/entities/bundle_entity.go:
```diff
@@ -140,7 +138,7 @@ func (b *BundleEntity) loadReplaces() error {
 	defer b.mu.Unlock()
 	if b.replaces == nil {
-		replaces, err := loadFromEntity[Replaces](b.Entity, "olm.replaces", optional)
+		replaces, err := loadFromEntity[Replaces](b.Entity, PropertyBundleReplaces, optional)
```
thank you!

m1kola commented:
Ok, here what I did:

* I added a new property into `CatalogdEntitySource`: which contains a name of the current bundle and (if present in catalog) name of the bundle it replaces. Currently in the `main` we only have replaces and no current bundle name.
* Updated `InstalledPackageVariableSource` and `Replaces` predicate to use the new prop
* Updated a bunch of tests (mostly `Entity` mocks to include new prop)

For now I'm trying to make minimal changes to fix the issue, but I have a feeling that things will change internall once we introduce [semver based upgrades](https://docs.google.com/document/d/1KIhV8oqr22Bp8lf370Aej2n-3uW_xGcTptONHHPRVYA/edit).

perdasilva commented on internal/resolution/entities/bundle_entity_test.go:
```diff
@@ -268,21 +268,21 @@ var _ = Describe("BundleEntity", func() {
 	Describe("Replaces", func() {
 		It("should return the replaces property if present", func() {
```
nit: should we rephrase the test description?

perdasilva replied:
```suggestion
		It("should return the channel entry property if present", func() {
```

perdasilva commented on internal/resolution/entities/bundle_entity_test.go:
```diff
@@ -268,21 +268,21 @@ var _ = Describe("BundleEntity", func() {
 	Describe("Replaces", func() {
```
```suggestion
	Describe("ChannelEntry", func() {
```

perdasilva reviewed (approved):
lgtm - thanks for this!! We're having some discussions around channels atm, so it _could_ change, but getting this stable rn is super important!
